# RNAseq Pipeline

## Inputs
- `config.yml` — configuration (under dataset root)
- `samplesheet.csv` — sample metadata and FASTQ paths (under dataset root, can be generated by running script at first time)
- Raw FASTQ files (paths referenced in the samplesheet.csv)

## Example usage
```
script_dir=/mnt/f/GitHub/myScript/NGS/CutTag
dataset_dir=/mnt/f/workspace/2026-01-01_CutTag_dataset

# Enter dataset directory
cd $dataset_dir

# Run to generate samplesheet.csv
bash $script_dir/run_cuttag.sh config.yml

# Edit samplesheet.csv
# Edit config.yml

# Re-run pipeline
bash $script_dir/run_cuttag.sh config.yml
```

## Outputs
- QC reports and trimmed FASTQs
- Aligned BAMs or quasi-mapping outputs
- Gene/transcript-level count tables
- logs

## Notes
- For complicated genomic setting, refer to `helper.sh`